subsys/power/device_pm.c

```diff
@@ -0,0 +1,239 @@
+ */
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
+					struct device_pm, dwork);
+	struct device *dev = pm->dev;
+	int ret;
+	u32_t request;
+
+	while (1) {
+		k_sem_take(&dev->pm->lock, K_FOREVER);
+		if (dev->pm->async_req_pending) {
+			request = dev->pm->async_request;
+			dev->pm->async_req_pending = false;
+		} else {
+			k_sem_give(&dev->pm->lock);
+			break;
+		}
+		k_sem_give(&dev->pm->lock);
+
+		ret = device_pm_state_handler(dev, request);
+	}
+}
+
+static int device_pm_request(struct device *dev,
+			     u32_t target_state, u32_t pm_flags)
+{
+	int ret = 0;
+
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		if (atomic_inc(&dev->pm->usage) < 0) {
+			goto req_out;
+		}
+	} else {
+		if (atomic_dec(&dev->pm->usage) > 1) {
+			goto req_out;
+		}
+	}
+
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	if (!dev->pm->enable) {
+		k_sem_give(&dev->pm->lock);
+		goto req_out;
+	}
+
+	/* Cancel the already queued work */
+	k_delayed_work_cancel(&dev->pm->dwork);
+
+	if (pm_flags & DEVICE_PM_ASYNC) {
+		dev->pm->async_request = target_state;
+		dev->pm->async_req_pending = true;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+					&dev->pm->dwork, dev->pm->timeout);
+	}
+	k_sem_give(&dev->pm->lock);
+
+	if (!(pm_flags & DEVICE_PM_ASYNC)) {
+		ret = device_pm_state_handler(dev, target_state);
+	}
+
+req_out:
+	return ret;
+}
+
+int device_pm_get(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_ACTIVE_STATE, DEVICE_PM_ASYNC);
+}
+
+int device_pm_get_sync(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_ACTIVE_STATE, 0);
+}
+
+int device_pm_put(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_SUSPEND_STATE, DEVICE_PM_ASYNC);
+}
+
+int device_pm_put_sync(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_SUSPEND_STATE, 0);
+}
+
+void device_pm_enable(struct device *dev)
+{
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->enable = true;
+
+	/* During the driver init, device can set the
+	 * PM state accordingly. For later cases we need
+	 * to check the usage and set the device PM state.
+	 */
+	if (!dev->pm->dwork.work.handler) {
+		dev->pm->dev = dev;
+		k_delayed_work_init(&dev->pm->dwork, pm_work_handler);
+		k_poll_signal_init(&dev->pm->signal);
+	} else {
+		k_delayed_work_cancel(&dev->pm->dwork);
+		if (dev->pm->usage) {
+			dev->pm->async_request = DEVICE_PM_ACTIVE_STATE;
+		} else {
+			dev->pm->async_request = DEVICE_PM_SUSPEND_STATE;
+		}
+		dev->pm->async_req_pending = true;
+		k_delayed_work_submit_to_queue(&devpm_workq,
+						&dev->pm->dwork, 0);
+	}
+	k_sem_give(&dev->pm->lock);
+}
+
+void device_pm_disable(struct device *dev)
+{
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	k_delayed_work_cancel(&dev->pm->dwork);
+	/* Bring up the device before disabling the Idle PM */
+	dev->pm->async_request = DEVICE_PM_ACTIVE_STATE;
+	dev->pm->async_req_pending = true;
+	k_delayed_work_submit_to_queue(&devpm_workq, &dev->pm->dwork, 0);
```

Comment:
The worker should be executed from the thread defined in this file and also cancel/submit requests are taken care by checking the async request pending field.

Reply:
why not using system workqueue? if you need to sleep on poll you can poll with no wait and if nor ready resubmit work with delay (should save ram)
cancel is not needed if you resubmit btw

Reply:
some time ago I have created an issue with proposal that work could be submitted when polled object is ready (https://github.com/zephyrproject-rtos/zephyr/issues/10748)

Reply:
> why not using system workqueue?

The device_set_power_state() is synchronous and it could take 1s to change power state.
If we will use system workqueue, we will block system while executing state change.

IMHO we need asynchronous interface to device_set_power_state as a prerequisite to this PR.

Reply:
@pizi-nordic I agree. It seems that this PR is trying to address the limitation of the model in the drivers. First we should fix the API there otherwise the solution will never be right.

Reply:
> > why not using system workqueue?
> 
> The device_set_power_state() is synchronous and it could take 1s to change power state.
> If we will use system workqueue, we will block system while executing state change.
> 
> IMHO we need asynchronous interface to device_set_power_state as a prerequisite to this PR.

@pizi-nordic how can we achieve that? do we need to introduce new API's? what synchronous device_st_power_state requests? Ideally, we would need both right?